contrib/hualloc/hu_alloc.h

```diff
@@ -0,0 +1,1615 @@
+        isHuge = true;
+    } else {
+        // regular alloc
+        sz = (sz + PAGE_SIZE - 1) & ~(PAGE_SIZE - 1);
+    }
+    void *ptr = mmap(0, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+    if (isHuge) {
+#ifdef MADV_HUGEPAGE
+        madvise(ptr, sz, MADV_HUGEPAGE);
+#endif
+    }
+    Y_ASSERT(ptr);
+    return ptr;
+}
+static void OsFree(void *p, yint sz)
+{
+    munmap(p, sz);
+}
+#endif
+
+static void* OsAlloc(yint szArg)
+{
+    yint sz = szArg;
+    return OsAlloc(&sz);
+}
+
+// smallest power of 2 greater or equal to val
+static yint CalcLog2(yint val)
+{
+    if (val <= 1) {
+        return 0;
+    }
+    unsigned long xx;
+    _BitScanReverse64(&xx, val - 1);
+    return xx + 1;
+}
+
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+static std::atomic<ui64> AllocatorIsInitialized;
+static void hu_init();
+static void hu_check_init()
+{
+    if (AllocatorIsInitialized != 1) {
+        hu_init();
+    }
+}
+
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////
+const yint GLOBAL_HASH_SIZE_LN = 20; // 4x+ number of blocks and more to avoid excessive hash lookups
+const yint GLOBAL_HASH_SIZE = 1ll << GLOBAL_HASH_SIZE_LN;
+
+struct TGAEntry
+{
+    std::atomic<ui64> Ptr;
+    yint Size;
+};
+static TGAEntry *GAHash;
+
+
+static void GAInit()
+{
+    GAHash = (TGAEntry*)OsAlloc(GLOBAL_HASH_SIZE * sizeof(TGAEntry));
+}
+
+
+static yint GAHashPtr(const void *ptr)
+{
+    ui64 x = (const char*)(ptr) - (const char*)nullptr;
+    ui64 res = (0x4847badfea31337ull * x) >> (64 - GLOBAL_HASH_SIZE_LN);
+    Y_ASSERT(res < GLOBAL_HASH_SIZE);
+    return res;
+}
+
+
+NOINLINE static void* GAAlloc(yint szArg)
+{
+    if (szArg == 0) {
+        return 0;
+    }
+    hu_check_init();
+    yint sz = szArg;
+    void *ptr = OsAlloc(&sz);
+    if (ptr == 0) {
+        abort(); // out of memory
+    }
+    yint k = GAHashPtr(ptr);
+    for (yint kStart = k;;) {
+        TGAEntry &ga = GAHash[k];
+        ui64 oldPtr = ga.Ptr.load();
+        if (oldPtr <= 1) {
+            ui64 newPtr = (ui64)ptr;
+            if (ga.Ptr.compare_exchange_weak(oldPtr, newPtr)) {
+                ga.Size = sz;
+                return ptr;
+            }
+            continue;
+        }
+        k = (k + 1) & (GLOBAL_HASH_SIZE - 1);
+        if (k == kStart) {
+            break;
+        }
+    }
+    abort(); // out of hash
+    return 0;
+}
+
+
+static void GAFree(void *ptr)
+{
+    if (ptr == 0) {
+        return;
+    }
+    yint k = GAHashPtr(ptr);
+    yint sz = 0;
+    for (;;) {
+        TGAEntry &ga = GAHash[k];
+        void *gaPtr = (void*)ga.Ptr.load();
+        if (gaPtr == ptr) {
+            sz = ga.Size;
+            ga.Ptr = 1;
+            break;
+        }
+        if (gaPtr == 0) {
+            abort(); // not found
+        }
+        k = (k + 1) & (GLOBAL_HASH_SIZE - 1);
+    }
```

Comment:
Here we don't check for `(k == kStart)`, making the loop endless when table is full and incorrect address is provided to GAFree

Reply:
calling free() for incorrect address can have arbitrary behavior?